### ISSUE TITLE
DbUnitExtension and transactional lifecycle issues(SpringExtension + TransactionalTestExecutionListener)

### ISSUE THREAD

**Issue by smaxx** [database-rider/database-rider#544]
Let's take a look at the following example. 

```
    @Test
    @ExpectedDataSet(value = "comparison/repository/pvsBlobsRepository/pvsBlobSaveExpected.yaml")
    @Commit
    void shouldSavePvsBlobs(){
        final PvsBlobs pvsBlobs = new PvsBlobs();
        pvsBlobs.setClaimId(CLAIM_ID);
        pvsBlobs.setDiff(DIFF);
        pvsBlobs.setTimestamp(TIMESTAMP);
        pvsBlobs.setCorrelationId(CORRELATION_ID);
        pvsBlobs.setTenantId(TENANT_ID);
        pvsBlobsRepository.save(pvsBlobs);
    }
```
Test class is declared as:
```
@DataJpaTest
@ContextConfiguration(classes = ComparisonConfiguration.class)
@EnableAutoConfiguration
@DBRider
@DBUnit(cacheConnection = false)
@ActiveProfiles("test")
abstract class BaseRepositoryTest {
```

The issue is that DbRider/DbUnit always says that there are no records inserted into a table.
After some analysis it appears that the root cause is wrong(?) usage of test lifecycle phases/callback methods.

What would be the correct sequence for this test?

1. Data is persisted in the test
2. Transaction is committed
3. DbRider/DbUnit verifies the DB state against expectations

What happens in real life is the opposite. DBUnitExtension has its assertion in "afterTestExecution" method while actual transaction is committed by SpringExtension#afterEach which in turn invokes all TestExecutionListener-s and their "afterTestMethod" callbacks incl TransactionalTestExecutionListener.
And just to remind you the sequence of lifecycle method executions(taken from Junit5 TestMethodTestDescriptor):
```
invokeBeforeEachCallbacks(context);
			if (throwableCollector.isEmpty()) {
				invokeBeforeEachMethods(context);
				if (throwableCollector.isEmpty()) {
					invokeBeforeTestExecutionCallbacks(context);
					if (throwableCollector.isEmpty()) {
						invokeTestMethod(context, dynamicTestExecutor);
					}
					invokeAfterTestExecutionCallbacks(context);
				}
				invokeAfterEachMethods(context);
			}
		invokeAfterEachCallbacks(context);
```

As a temporary workaround i had to start using custom DbUnitExtension subclass and it fixes the issue:
```
public class CustomDbUnitExtension extends DBUnitExtension {

    @Override
    public void afterTestExecution(final ExtensionContext extensionContext) throws Exception {
        // Do nothing here
    }

    @Override
    public void afterEach(final ExtensionContext extensionContext) throws Exception {
        super.afterTestExecution(extensionContext);
        super.afterEach(extensionContext);
    }
}
```

**Comment by TobiasBerndt** [comment]
You can also try to override the `@Transactional` from `@DataJpaTest` by suspending the transactions of test method.
```
@DataJpaTest
@Transactional(propagation = Propagation.NOT_SUPPORTED) // Execute non-transactionally, suspend the current transaction if one exists
@ContextConfiguration(classes = ComparisonConfiguration.class)
@EnableAutoConfiguration
@DBRider
@DBUnit(cacheConnection = false)
@ActiveProfiles("test")
abstract class BaseRepositoryTest {
```

**Comment by smaxx** [comment]
True, it is an option for sure. But the point is that we live in a "Spring world", and assuming that its infrastructure uses the aforementioned annotations(DataJpaTest, DataJdbcTest, JdbcTest and the others), all of which have in common the same part, all of those declare the test class/method `@Transactional`. And in turn the annotation is handled by Spring via SprinExtension.
So what i wanted to say is:

- there is a widely used approach for testing Spring(Boot)-based applications
- it doesn't work for some reason in conjunction with DbRider
- someone has to adjust the code, otherwise only workarounds(changing the verification phase or turning off transactionality of the test method) can do the trick
- taking into account scale of Spring vs DbRider i definitely think that the changes have to be done here, in DBUnitExtension

**Comment by rmpestano** [comment]
Hey folks, we discussed this [before here](https://github.com/database-rider/database-rider/issues/183#issuecomment-625678128) and at the time we agreed that having transactional tests in dbrider doesn't make much sense so we decided not to support it. If you don't agree and have a solution I'm happy to discuss and review it and depending on how complex it is we can add to rider-junit5

**Comment by darkolakovic** [comment]
Hi, unfortunately not participating in spring transactions is causing a lot of issues in different setups based on different parameters. For instance, having auto-commit set to true is causing performance issues when there are a lot of tests (deletes of tables can take an hour for 500+ tests), and setting it to false is causing visibility issues (queries sometimes do not see inserted dataset and also expected datasets asserts are failing).  

In order to participate in a transaction, connection should be obtained through `DataSourceUtils.getConnection(dataSource) `instead of `dataSource.getConnection()`.
Perhaps `getConnectionHolder` method could also accept a function that accepts `DataSource` and returns `Connection`, so that it can be adjusted for different cases.
